source/common/common/cancel_wrapper.h

```diff
@@ -0,0 +1,41 @@
+#pragma once
+
+#include <memory>
+#include <utility>
+
+#include "absl/functional/any_invocable.h"
+
+namespace Envoy {
+namespace CancelWrapper {
+
+using CancelFunction = absl::AnyInvocable<void()>;
+
+// Wraps a callback with a cancellation function. The cancellation function is saved
+// in an 'out' pointer argument to facilitate simple chaining in place, e.g. typical
+// usage would be of the form
+//
+// someFunctionThatTakesACallback(cancelWrapped([this](SomeResponse response) {
+//   doStuff(response);
+// }, &cancel_callback_));
+//
+// Then cancel_callback_ could be called (conditionally on having been populated) in the
+// class destructor to prevent use of "this" in a dispatched event that ends up occurring
+// after the calling object was destroyed.
+//
+// This cancellation is not safe to be called between threads - it is intended to be used in
```

Comment:
To help detect if this happens, wdyt about, in debug builds only, calling `std::this_thread::get_id()` and storing it, and comparing to that when the cancellation is called?

Reply:
My thought was if it's used across threads it would trigger a tsan failure so would provoke the same thing anyway in tests, and I don't think debug builds are common outside of tests.
(And also it's going to be replacing code that already does the same thing without that extra guard.)

Reply:
TSAN will only detect it if operations are happening on multiple threads in a short time window, which may or may not happen in the tests

Reply:
I just had a go at doing this and getting it to verify the expected thing basically involves having totally separate debug and non-debug versions of the function (because both the lambdas need to capture the thread id and then do something with it at which point you either have 4 ifdefs and only 2 or 3 common lines, or one ifdef and repeat everything). And once you have that it seems like what's even the point, now you've got twice as much code to maintain for what was supposed to be a simple helper function. I aborted it.

Reply:
test/test_common/only_one_thread.h may be a good pattern.

Reply:
If that was available as a thing wrapped in `#ifndef` rather than as a test library then that would almost work, except it would still need its object to be captured into the lambdas which would mean those lambda declarations would also need to be inside ifdef blocks so it would still end up basically the same mess of spaghetti as what I gave up on.

However, I'm now thinking maybe it *would* be worth having just the whole two-implementations thing going on for this, because some of the cases I'm replacing did previously have asserts for same-threadedness (previously using dispatcher) and it makes sense to offload those checks into the common code too.